File: backend/app/routers/roles.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from supabase import Client

from ..deps import get_current_user_id, get_user_supabase_client
from ..schemas import RoleCreate, RoleOut, RoleUpdate

router = APIRouter(prefix="/roles", tags=["roles"])


@router.get("", response_model=list[RoleOut])
def list_roles(client: Client = Depends(get_user_supabase_client)) -> list[dict]:
    response = (
        client.table("roles").select("*").order("start_date", desc=True).execute()
    )
    return response.data


@router.post("", response_model=RoleOut, status_code=status.HTTP_201_CREATED)
def create_role(
    role: RoleCreate,
    client: Client = Depends(get_user_supabase_client),
    user_id: str = Depends(get_current_user_id),
) -> dict:
    payload = role.model_dump(mode="json") | {"user_id": user_id}
    response = client.table("roles").insert(payload).execute()
    return response.data[0]


@router.get("/{role_id}", response_model=RoleOut)
def get_role(role_id: UUID, client: Client = Depends(get_user_supabase_client)) -> dict:
    response = client.table("roles").select("*").eq("id", str(role_id)).execute()
    if not response.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Role not found")
    return response.data[0]


@router.patch("/{role_id}", response_model=RoleOut)
def update_role(
    role_id: UUID,
    role: RoleUpdate,
    client: Client = Depends(get_user_supabase_client),
) -> dict:
    payload = role.model_dump(mode="json", exclude_unset=True)
    if not payload:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "No fields to update")
    response = client.table("roles").update(payload).eq("id", str(role_id)).execute()
    if not response.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Role not found")
    return response.data[0]


@router.delete("/{role_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_role(role_id: UUID, client: Client = Depends(get_user_supabase_client)) -> None:
    client.table("roles").delete().eq("id", str(role_id)).execute()

File: backend/src/routes/roles.ts
import { Router } from "express";
import { requireAuth, type AuthedRequest } from "../middleware/auth.js";
import { validateBody } from "../middleware/validate.js";
import { roleCreate, roleUpdate } from "../schemas/role.js";
import { getUserSupabaseClient } from "../supabaseClients.js";

export const rolesRouter = Router();

rolesRouter.use(requireAuth);

rolesRouter.get("", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  const { data, error } = await client
    .from("roles")
    .select("*")
    .order("start_date", { ascending: false });
  if (error) {
    res.status(400).json({ detail: error.message });
    return;
  }
  console.log("data in roles.ts", data);
  res.json(data);
});

rolesRouter.post("", validateBody(roleCreate), async (req, res) => {
  const { token, userId } = req as AuthedRequest;
  const client = getUserSupabaseClient(token);
  const payload = { ...req.body, user_id: userId };
  console.log("payload", payload);
  const { data, error } = await client.from("roles").insert(payload).select().single();
  console.log("data", data);
  console.log("error", error);
  if (error) {
    res.status(400).json({ detail: error.message });
    return;
  }
  res.status(201).json(data);
});

rolesRouter.get("/:roleId", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  const { data, error } = await client
    .from("roles")
    .select("*")
    .eq("id", req.params.roleId);
  if (error || !data?.length) {
    res.status(404).json({ detail: "Role not found" });
    return;
  }
  res.json(data[0]);
});

rolesRouter.patch("/:roleId", validateBody(roleUpdate), async (req, res) => {
  if (Object.keys(req.body).length === 0) {
    res.status(400).json({ detail: "No fields to update" });
    return;
  }
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  const { data, error } = await client
    .from("roles")
    .update(req.body)
    .eq("id", req.params.roleId)
    .select();
  if (error || !data?.length) {
    res.status(404).json({ detail: "Role not found" });
    return;
  }
  res.json(data[0]);
});

rolesRouter.delete("/:roleId", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  await client.from("roles").delete().eq("id", req.params.roleId);
  res.status(204).send();
});
